clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -698,36 +777,54 @@ synchronized boolean hasUnresolvedSequences() {
     }
 
     synchronized boolean hasUnresolvedSequence(TopicPartition topicPartition) {
-        return partitionsWithUnresolvedSequences.contains(topicPartition);
+        return partitionsWithUnresolvedSequences.containsKey(topicPartition);
     }
 
-    synchronized void markSequenceUnresolved(TopicPartition topicPartition) {
-        log.debug("Marking partition {} unresolved", topicPartition);
-        partitionsWithUnresolvedSequences.add(topicPartition);
+    synchronized void markSequenceUnresolved(ProducerBatch batch) {
+        int nextSequence = batch.lastSequence() + 1;
+        log.debug("Marking partition {} unresolved with next sequence number {}", batch.topicPartition, nextSequence);
+        partitionsWithUnresolvedSequences.put(batch.topicPartition, nextSequence);
```

Comment:
Potentially, if multiple batches time out in the accumulator. In that case I think we would only care about the sequence number of the highest-sequence one, since we use that to definitively determine that there is a gap in the log. I'll change this to only update the value if the sequence is greater than the existing one.